care/emr/api/viewsets/diagnostic_report.py

```diff
@@ -198,4 +203,79 @@ def upsert_observations(self, request, *args, **kwargs):
+                age = date_delta.days / 365.25
+
+                context = {
+                    "gender": patient.gender,
+                    "age": age,
+                    "applies_to": [
+                        tag.get("slug")
+                        for tag in PatientInstanceTagManager().render_tags(patient)
+                    ],
+                }
+
+                def extract_value(val: Any) -> Any:
+                    """Unified value extraction matching value_fits logic."""
+                    if isinstance(val, dict):
+                        if "coding" in val and val["coding"] is not None:
+                            return val["coding"].get("code")
+                        if "quantity" in val:
+                            return val["quantity"]
+                        if "value" in val:
+                            return val["value"]
+                    return val
+
+                try:
+                    if not model_instance.component:
+                        engine = InterpretationEvaluator(
+                            model_instance.observation_definition.qualified_ranges
+                        )
+                        value_to_evaluate = extract_value(model_instance.value)
+                        model_instance.interpretation = engine.evaluate(
+                            context, value_to_evaluate
+                        )
+                        matched_condition = engine.get_matching_condition(
+                            context=context
+                        )
+                        if matched_condition:
+                            model_instance.reference_range = matched_condition.get(
+                                "ranges", []
+                            )
+                    else:
+                        component_definition_dict = {
+                            component_def["code"]["code"]: component_def[
+                                "qualified_ranges"
+                            ]
+                            for component_def in model_instance.observation_definition.component
+                        }
+                        for component in model_instance.component:
+                            component_code = component.get("code", {}).get("code")
+                            component_value = extract_value(component.get("value"))
+                            engine = InterpretationEvaluator(
+                                component_definition_dict.get(component_code, [])
+                            )
+                            component["interpretation"] = engine.evaluate(
+                                context, component_value
+                            )
+
+                            matched_condition = engine.get_matching_condition(
+                                context=context
+                            )
+                            if matched_condition:
+                                model_instance.reference_range = matched_condition.get(
```

Comment:
is this assignment correct?
model_instance.reference_range is being reassigned in the loop
if we are okay with the first correct assignment then we can break out of the loop

also this assignment is defaulting to [] as compared to {} on line 248